internal/k8s/controllers/configuration_state_controller.go

```diff
@@ -0,0 +1,125 @@
+// SPDX-License-Identifier:Apache-2.0
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	metallbv1beta1 "go.universe.tf/metallb/api/v1beta1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+type ConfigurationStateReconciler struct {
+	client.Client
+	Logger         log.Logger
+	Scheme         *runtime.Scheme
+	ConfigStateRef *metallbv1beta1.ConfigurationState
+}
+
+func (r *ConfigurationStateReconciler) String() string {
+	return "ConfigurationStateReconciler"
+}
+
+func (r *ConfigurationStateReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	level.Info(r.Logger).Log("controller", r, "start reconcile", req.String())
+	defer level.Info(r.Logger).Log("controller", r, "end reconcile", req.String())
+
+	var configStatus metallbv1beta1.ConfigurationState
+	if err := r.Get(ctx, req.NamespacedName, &configStatus); err != nil {
+		if !apierrors.IsNotFound(err) {
+			level.Error(r.Logger).Log("controller", r, "error", "failed to get ConfigurationState", "error", err)
+			return ctrl.Result{}, err
+		}
+
+		if err := r.Create(ctx, r.ConfigStateRef); err != nil {
+			level.Error(r.Logger).Log("controller", r, "error", "failed to create ConfigurationState", "error", err)
+			return ctrl.Result{}, err
+		}
+
+		level.Info(r.Logger).Log("controller", r, "error", "ConfigurationState created", "name", r.ConfigStateRef.Name)
+		return ctrl.Result{}, nil
+	}
+
+	result := metallbv1beta1.ConfigurationStateResultUnknown
+	lastError := ""
+
+	// Only aggregate if there are conditions reported
+	if len(configStatus.Status.Conditions) > 0 {
+		result = metallbv1beta1.ConfigurationStateResultValid
+
+		var msgs []string
+		for _, cond := range configStatus.Status.Conditions {
+			if cond.Status == metav1.ConditionFalse {
```

Comment:
this assumes that status = false is an error condition, which might change depending on the conditions we are adding. Would it make sense to have a "isConditionErrorType" filter function over the type where we check the current conditions we have? If the filter fails we continue